### PR TITLE
[SPARK-20749][SQL] Built-in SQL Function Support - all variants of LEN[GTH]

### PR DESCRIPTION
**Pull request by kiszk**
## What changes were proposed in this pull request?

This PR adds built-in SQL function `BIT_LENGTH()`, `CHAR_LENGTH()`, and `OCTET_LENGTH()` functions.

`BIT_LENGTH()` returns the bit length of the given string or binary expression.
`CHAR_LENGTH()` returns the length of the given string or binary expression. (i.e. equal to `LENGTH()`)
`OCTET_LENGTH()` returns the byte length of the given string or binary expression.

## How was this patch tested?

Added new test suites for these three functions

### CODE REVIEW

**Comment by maropu** [comment]
It seems you have the wrong JIRA number. Also, you need to add tests in `SQLQueryTestSuite`. Thanks.

**Comment by kiszk** [comment]
@maropu Thank you for pointing them out. I addressed both.

**Comment by kiszk** [comment]
Jenkins, retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 1]
```diff
@@ -1180,6 +1180,60 @@ case class Length(child: Expression) extends UnaryExpression with ImplicitCastIn
...
 
 /**
+ * A function that return the bit length of the given string or binary expression.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(expr) - Returns the bit length of `expr` or number of bits in binary data.",
+  extended = """
+    Examples:
+      > SELECT _FUNC_('Spark SQL');
+       9
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: Is it 9 bits?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 2]
```diff
@@ -277,6 +277,8 @@ object FunctionRegistry {
     // string functions
     expression[Ascii]("ascii"),
     expression[Base64]("base64"),
+    expression[BitLength]("bit_length"),
+    expression[Length]("char_length"),
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: Is it common that `length` function is just alias of `char_length`?

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: E.g., in mysql, `length` is an alias of `octet_length`.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: Seems `length` is an length of `char_length` too in Hive.

Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: I think that `length` is an alias of `octet_length` in mysql. In postgreSql, `length` is alias an of `char_length`.

File: sql/core/src/test/resources/sql-tests/results/operators.sql.out [hunk 3]
```diff
@@ -1,5 +1,5 @@
 -- Automatically generated by SQLQueryTestSuite
--- Number of queries: 38
+-- Number of queries: 41
```
Comment (viirya) [inline sql/core/src/test/resources/sql-tests/results/operators.sql.out]: Seems you forget to update to `operators.sql` too?

File: sql/core/src/test/resources/sql-tests/results/operators.sql.out [hunk 4]
```diff
@@ -316,3 +316,25 @@ select cot(-1)
...
+
+-- !query 39
+select character_length(`abc`)
+-- !query 39 schema
+struct<CHARACTER_LENGTH(abc):int>
+-- !query 39 output
+3
+
+-- !query 40
+select octet_length(`abc`)
```
Comment (viirya) [inline sql/core/src/test/resources/sql-tests/results/operators.sql.out]: Should we use an example string which returns different results on `character_length` and `octet_length`?

Comment (kiszk) [inline sql/core/src/test/resources/sql-tests/results/operators.sql.out]: [These test cases](https://github.com/apache/spark/pull/18046/files#diff-e4663e57952b37150642b33b998715a8R527) handles the case.

File: sql/core/src/test/resources/sql-tests/results/operators.sql.out (hunk 3 above)
Comment (kiszk) [inline sql/core/src/test/resources/sql-tests/results/operators.sql.out]: Thanks, added.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 1 above)
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: Good catch. updated.

**Comment by viirya** [comment]
Can we also update the doc of `Length`? It is better to explicitly say it returns character length.

**Comment by kiszk** [comment]
@viirya I agree with you. However, I leave it for consistency of documents among versions. 
I would like to hear @gatorsmile 's opinion.

**Comment by kiszk** [comment]
Jenkins, retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: Based on the ANSI SQL, `LENGTH` function as a synonym for `CHAR_LENGTH`.

**Comment by gatorsmile** [comment]
@kiszk @viirya The document of `length` needs an update. Please always correct the function description if you find anything is incorrect. Thanks!

**Comment by gatorsmile** [comment]
Please run the following command to generate the result file:
```
SPARK_GENERATE_GOLDEN_FILES=1 build/sbt "sql/test-only *SQLQueryTestSuite"
```

**Comment by kiszk** [comment]
@gatorsmile thank you for your suggestion.

**Comment by kiszk** [comment]
@gatorsmile Do I have to do additional things for this PR?

**Comment by gatorsmile** [comment]
@kiszk Could you resolve the conflict? Thanks!

**Comment by kiszk** [comment]
Jenkins, retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 5]
```diff
@@ -1153,15 +1153,17 @@ case class Substring(str: Expression, pos: Expression, len: Expression)
 }
 
 /**
- * A function that return the length of the given string or binary expression.
+ * A function that return the char length of the given string or binary expression.
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: ... or number of bytes of binary expression.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 6]
```diff
@@ -1180,6 +1182,60 @@ case class Length(child: Expression) extends UnaryExpression with ImplicitCastIn
 }
 
 /**
+ * A function that return the bit length of the given string or binary expression.
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: ... that returns ...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 7]
```diff
@@ -1180,6 +1182,60 @@ case class Length(child: Expression) extends UnaryExpression with ImplicitCastIn
...
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    child.dataType match {
+      case StringType => defineCodeGen(ctx, ev, c => s"($c).numBytes() * 8")
+      case BinaryType => defineCodeGen(ctx, ev, c => s"($c).length * 8")
+    }
+  }
+}
+
+/**
+ * A function that return the byte length of the given string or binary expression.
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: ditto

**Comment by kiszk** [comment]
ping @gatorsmile

**Comment by kiszk** [comment]
ping @gatorsmile

**Comment by kiszk** [comment]
ping @gatorsmile

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
@kiszk Sorry, I missed this ping.

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Pull request closed without merging**